### PR TITLE
[huggingface_pytorch] Add HuggingFace TGI dockerfile

### PR DESCRIPTION
**Pull request by xyang16**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description

We want to release Huggingface TGI DLC, which include the Text Generation Inference (https://github.com/huggingface/text-generation-inference) for large language model inference.

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+# Install router
+COPY --from=builder /usr/src/target/release/text-generation-router /usr/local/bin/text-generation-router
+# Install launcher
+COPY --from=builder /usr/src/target/release/text-generation-launcher /usr/local/bin/text-generation-launcher
+
+# AWS Sagemaker compatible image
+FROM base as sagemaker
+
+COPY --chmod=775 sagemaker-entrypoint.sh entrypoint.sh
```
Comment (philschmid) [inline huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:86]: Where do we place the sagemaker-entrypoint.sh?

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+RUN cargo chef cook --release --recipe-path recipe.json
+
+COPY Cargo.toml Cargo.toml
+COPY rust-toolchain.toml rust-toolchain.toml
+COPY proto proto
+COPY router router
+COPY launcher launcher
+RUN cargo build --release
+
+FROM nvidia/cuda:11.8.0-devel-ubuntu22.04 as base
```
Comment (philschmid) [inline huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: We are not using an nvidia image directly see: https://github.com/huggingface/text-generation-inference/blob/e86cca9723d8f7b9a17f11d716240d0ad3e0f875/Dockerfile#L35

Comment (philschmid) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: this is needed for flash attention cuda kernels: https://github.com/huggingface/text-generation-inference/blob/e86cca9723d8f7b9a17f11d716240d0ad3e0f875/Dockerfile#L77

https://github.com/huggingface/text-generation-inference/blob/e86cca9723d8f7b9a17f11d716240d0ad3e0f875/Dockerfile#L88

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    NUM_SHARD=1 \
+    PORT=80 \
+    CUDA_HOME=/usr/local/cuda \
+    LD_LIBRARY_PATH="/opt/miniconda/envs/text-generation/lib:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:$LD_LIBRARY_PATH" \
+    CONDA_DEFAULT_ENV=text-generation \
+    PATH=$PATH:/opt/miniconda/envs/text-generation/bin:/opt/miniconda/bin:/usr/local/cuda/bin
+
+RUN apt-get update && apt-get install -y git curl libssl-dev unzip && rm -rf /var/lib/apt/lists/*
+
+RUN cd ~ && \
```
Comment (philschmid) [inline huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:48]: We also moved to mamba: https://github.com/huggingface/text-generation-inference/blob/e86cca9723d8f7b9a17f11d716240d0ad3e0f875/Dockerfile#L58

**Review by philschmid — COMMENTED** [review]
I left some comments, we made several improvements in the last weeks to remove conda and use mamba as well how build and add custom cuda kernels. See for the latest version: https://github.com/huggingface/text-generation-inference/blob/main/Dockerfile

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu (hunk 2 above)
Comment (lanking520) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: @philschmid do you have and source code lead to the base image you are using? We need to verify the license and agreement holds by these container owners

Comment (lanking520) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: The reason we are using NVIDIA-CUDA image is most of them is cleared from OSS and legal

Comment (xyang16) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: > We are not using an nvidia image directly see: https://github.com/huggingface/text-generation-inference/blob/e86cca9723d8f7b9a17f11d716240d0ad3e0f875/Dockerfile#L35

I was referring to 0.5.0 Dockerfile, which is using nvidia/cuda:11.8.0-devel-ubuntu22.04 image as base: https://github.com/huggingface/text-generation-inference/blob/v0.5.0/Dockerfile#L30

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu (hunk 1 above)
Comment (xyang16) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:86]: In WORKDIR /usr/src. This is the same with TGI 0.5.0 Dockerfile: https://github.com/huggingface/text-generation-inference/blob/v0.5.0/Dockerfile#L30

File: huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu (hunk 2 above)
Comment (philschmid) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: The `pytorch-install` base image is just plain `debian:bullseye-slim`

Comment (philschmid) [reply huggingface/pytorch/tgi/docker/0.5/py3/cu118/Dockerfile.gpu:30]: The only base images we use are `lukemathwalker/cargo-chef:latest-rust-1.69` and `debian:bullseye-slim`

**Pull request closed without merging**